### PR TITLE
tock: fix lowrisc/opentitan uart hw driver

### PR DESCRIPTION
**Pull request by ccteng-google**
The lowrisc/opentitan uart hw driver was not sending
tx_client.transmitted_buffer notification properly causing
app (user mode) console write syscall to never receive the
write complete subscription callback.

Tested with opentitan earlgrey running Tock OS in verilator
simulation and app calling console write to verify successful
return from syscall.

Change-Id: Id1e485373466d6aa414cc011b05b4b5c559e17e0

### Pull Request Overview

This pull request fixes the problem where 
the lowrisc/opentitan uart hw driver was not sending
tx_client.transmitted_buffer notification properly causing
Tock OS app (user mode) console write syscall to never
receive the write complete subscription callback.

### Testing Strategy

This pull request was tested by
running opentitan earlgrey + Tock OS + app in verilator
simulation and app calling console write to verify successful
return from syscall

### TODO or Help Wanted

This pull request still needs...

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by lschuermann** [comment]
This looks very similar to an issue I had with the LiteX codebase. Unfortunately I don't think this fix is entirely valid, given that as far as I can tell this will potentially issue a callback on the first call to `transmit` and this can cause other issues down the line. This is further illustrated in the HIL TRD.

In this case we will probably want to schedule a deferred call. I can recommend you to take a look here:

https://github.com/tock/tock/blob/ba3baef1a677704cbf8e15ef72a4161c41df1c10/chips/litex/src/uart.rs#L386-L402

File: chips/lowrisc/src/uart.rs [hunk 1]
```diff
@@ -192,18 +192,26 @@ impl<'a> Uart<'a> {
...
                     regs.wdata.write(wdata::data.val(tx_buf[tx_idx] as u32));
                     self.tx_index.set(tx_idx + 1)
                 }
-            });
+                true
+            }) {
+                // We sent everything to the UART hardware
+                self.tx_client.map(|client| {
+                    self.tx_buffer.take().map(|tx_buf| {
+                        client.transmitted_buffer(tx_buf, self.tx_len.get(), Ok(()));
```
Comment (bradjc) [inline chips/lowrisc/src/uart.rs:211]: Unfortunately we can't call a callback here. If I'm reading this driver correctly, `transmit_buffer()` is called from hil::uart::transmit::transmit_buffer(), which means a callback could occur in response to a HIL command. Per Tock's callback model, callbacks can _only_ be triggered in response to interrupts.

**Comment by bradjc** [comment]
Beat me to it!

**Comment by lschuermann** [comment]
As discussed on today's call, this will probably have to be fixed in some Tock 2.0 release candidate, but it's fine to tag the initial one without. I can also offer to just implement it.

**Comment by ccteng-google** [comment]
@lschuermann Feel free to take it over and I'd be OK to close this PR.

**Comment by lschuermann** [comment]
Let's leave this open, I think I should be able to push to your branch. I'll test with Verilator, but I'd appreciate you or @alistair23  doing HW tests.

File: chips/lowrisc/src/uart.rs (hunk 1 above)
Comment (phil-levis) [reply chips/lowrisc/src/uart.rs:211]: Callbacks do not have to be triggered in response to interrupts; they just can't be invoked on the same call stack from the causing downcall. This generally means that it needs to either be from an interrupt or a deferred procedure call.

But yes, invoking `transmitted_buffer` here is incorrect.

**Comment by ccteng-google** [comment]
I will close this PR while I upgrade to latest opentitan for further investigation and reopen if needed.

**Comment by lschuermann** [comment]
Reopening given @alistair23 has confirmed he can reproduce this bug, even on older commits of Tock and/or OpenTitan/EarlGrey.

**Comment by phil-levis** [comment]
@alistair23 reports that his failure was not due to the UART, but due to memory allocation.

**Comment by alistair23** [comment]
Just to clarify.

I was seeing failures with some of the print release tests yesterday. I just assumed it was this issue and I then tried to see if I could find the last working version, which I wasn't able to find.

After looking into it today I realised that the problem is that the memory requirements (the `memop` commands) for the app were failing which is why it wasn't working.

So I am unable to reproduce any failures on the FPGA.

**Pull request closed without merging**